lib/raxx/session/signed_cookie.ex

```diff
@@ -0,0 +1,206 @@
+
+    %__MODULE__{
+      secret: secret,
+      cookie_name: cookie_name,
+      previous_secrets: previous_secrets
+    }
+  end
+
+  @doc """
+  Overwrite a clients session with a new value.
+  """
+  def embed(response, session, session_config = %__MODULE__{}) do
+    payload = safe_encode(session)
+    digest = safe_digest(payload, session_config.secret)
+
+    # Where to put 4096 check
+    # It could be in Raxx.set_cookie
+    response
+    |> Raxx.set_header(
+      "set-cookie",
+      SetCookie.serialize(session_config.cookie_name, payload <> "--" <> digest)
+    )
+  end
+
+  @doc """
+  Extract and verify the session sent from a client.
+  """
+  def extract(request, session_config = %__MODULE__{}) do
+    case Raxx.get_header(request, "cookie") do
+      nil ->
+        {:error, :no_cookies_sent}
+
+      cookie_header ->
+        case Cookie.parse(cookie_header) do
+          cookies = %{} ->
+            case Map.fetch(cookies, "#{session_config.cookie_name}") do
+              {:ok, session_cookie} ->
+                case String.split(session_cookie, "--", parts: 2) do
+                  [payload, digest] ->
+                    if verify_signature(payload, digest, [session_config.secret | session_config.previous_secrets]) do
+                      safe_decode(payload)
+                    else
+                      {:error, :could_not_verify_signature}
+                    end
+                  _ ->
+                    {:error, :invalid_session_cookie}
+                end
+              :error ->
+                {:error, :no_session_cookie}
+            end
+        end
+    end
+  end
+
+  @doc """
+  Sent a response to client informing it to clear a session.
+  """
+  def expire(response, session_config = %__MODULE__{}) do
+    response
+    |> Raxx.set_header("set-cookie", SetCookie.expire(session_config.cookie_name))
+  end
+
+  defp safe_digest(payload, secret) do
+    :crypto.hmac(:sha256, secret, payload)
+    |> Base.url_encode64()
+  end
+
+  defp safe_encode(term) do
+    {:ok, encoded_session} = encode(term)
+    Base.url_encode64(encoded_session)
+  end
+
+  defp encode(term) do
+    {:ok, :erlang.term_to_binary(term)}
+  end
+
+  defp safe_decode(binary) do
+    {:ok, encoded} = Base.url_decode64(binary)
+    decode(encoded)
+  end
+
+  # NOTE make sure decode is only called after verifying digest
+  # https://elixirforum.com/t/static-and-session-security-fixes-for-plug/3913
+  defp decode(binary) do
+    try do
+      term = :erlang.binary_to_term(binary)
+      {:ok, term}
+    rescue
+      _e in ArgumentError ->
+        {:error, :unable_to_decode_session}
+    end
+  end
+
+  defp verify_signature(payload, digest, secrets) do
+    Enum.any?(secrets, fn(secret) ->
+      secure_compare(digest, safe_digest(payload, secret))
+    end)
+  end
+
+  defp secure_compare(left, right) do
+    if byte_size(left) == byte_size(right) do
+      secure_compare(left, right, 0) == 0
+    else
+      false
```

Comment:
You're revealing information about the length of the token here.

Reply:
yeah I thought that. but it's whats in Plugs Crypto lib